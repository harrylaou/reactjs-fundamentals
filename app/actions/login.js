export const LOGIN_USER_OK = 'LOGIN_USER_OK'
export const LOGOUT_USER = 'LOGOUT_USER'

export const loginUser = () => ({
  type: LOGIN_USER_OK
})

export const logoutUser = () => ({
  type: LOGOUT_USER
})