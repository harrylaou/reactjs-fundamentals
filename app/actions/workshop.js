export const RECEIVE_WORKSHOP = 'RECEIVE_WORKSHOP'


export const receiveWorkshop = workshop => ({
  type: RECEIVE_WORKSHOP,
  workshop
})