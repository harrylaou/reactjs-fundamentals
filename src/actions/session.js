export const LOG_OUT = 'LOG_OUT'
export const LOGGED_IN = 'LOGGED_IN'

export const logout = () => ({
  type: LOG_OUT
})

export const loggedIn = token => ({
  type: LOGGED_IN,
  token
})
