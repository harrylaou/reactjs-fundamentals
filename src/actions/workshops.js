export const RECEIVE_WORKSHOPS = 'RECEIVE_WORKSHOPS'

export const receiveWorkshops = workshops => ({
  type: RECEIVE_WORKSHOPS,
  workshops
})
