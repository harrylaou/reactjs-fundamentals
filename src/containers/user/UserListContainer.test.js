import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import ConnectedUserListContainer, {UserListContainer} from './UserListContainer'
import { mockFetchContainer } from '../../testUtils'
import getMuiTheme from 'material-ui/styles/getMuiTheme'

import FakeConfigureStore from 'redux-mock-store'
import ActualConfigureStore from '../../store'
import {Provider} from 'react-redux'
import * as actions from '../../actions/users'

import injectTapEventPlugin from 'react-tap-event-plugin'
injectTapEventPlugin()

describe('<UserListContainer />', () => {
  it('should render with default props', () => {
    const push = jest.fn()
    const props = {
      params: { username: '123' },
      width: 3,
      router: {push},
      users: []
    }

    const wrapper = shallow(
      <UserListContainer {...props} />
    )

    expect(shallowToJson(wrapper)).toMatchSnapshot()

    wrapper.instance().showUserProfile({username: 'manolo'})

    expect(push).toBeCalledWith('/users/manolo')
  })


  const usersData = {
    users: [{
      name: {
        title: 'mr',
        first: 'Eduard',
        last: 'lama'
      },
      username: 'edu',
      email: 'eduard@gmail.com',
      gender: 'male'
    }]
  }

  const initialState = {
    users: [{
      name: {
        title: 'mr',
        first: 'Manolo',
        last: 'lama'
      },
      username: 'mano',
      email: 'manolo@gmail.com',
      gender: 'male'
    }]
  }
  const mockStore = FakeConfigureStore()

  const push = jest.fn()
  const replace = () => {}
  const go = () => {}
  const goBack = () => {}
  const goForward = () => {}
  const setRouteLeaveHook = () => {}
  const isActive = () => {}

  const props = {
    params: { username: '123' },
    width: 3,
    router: {
      push,
      replace,
      go,
      goBack,
      goForward,
      setRouteLeaveHook,
      isActive,
      setRouteLeaveHook
    },
    users: []
  }
  let store

  describe('<ConnectedUserListContainer />', () => {
    it('should render connected component', () => {
      store = mockStore(initialState)
      const container = shallow(
        <ConnectedUserListContainer
          {...props}
          store={store}
        />
      )

      expect(shallowToJson(container)).toMatchSnapshot()
    })
  })

  describe('Integration test: Mount + Wrapping Provider', () => {
    let wrapper

    beforeEach(()=>{
      store = mockStore(initialState)

      mockFetchContainer(usersData.users)

      wrapper = mount(
        <Provider store={store} >
          <ConnectedUserListContainer
            {...props}
          />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes:{
            muiTheme: React.PropTypes.object.isRequired,
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedUserListContainer).length).toEqual(1)
    })

    it('should Prop matches with initialState', () => {
      expect(wrapper.find(UserListContainer).prop('users')).toEqual(initialState.users)
    })

    it('should check action dispatch', () => {
      store.dispatch(actions.receiveUsers(usersData.users))
      const action = store.getActions()

      expect(action[0].type).toBe(actions.RECEIVE_USERS)
    })
  })

  describe('Integration test: Mount + Wrapping Provider + Actual Store + Reducers', () => {
    let wrapper

    beforeEach(()=>{
      store = ActualConfigureStore(initialState)

      mockFetchContainer(usersData.users)

      wrapper = mount(
        <Provider store={store} >
          <ConnectedUserListContainer
            {...props}
          />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes:{
            muiTheme: React.PropTypes.object.isRequired,
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedUserListContainer).length).toEqual(1)
    })

    it('should check action dispatch', () => {
      //OR store.dispatch(actions.receiveUsers(usersData))
      wrapper.find(UserListContainer).props().dispatch(actions.receiveUsers(usersData.users))

      expect(wrapper.find(UserListContainer).prop('users')).toEqual(usersData.users)
    })
  })
})