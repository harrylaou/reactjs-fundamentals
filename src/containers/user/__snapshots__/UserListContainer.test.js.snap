// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserListContainer /> <ConnectedUserListContainer /> should render connected component 1`] = `
<withRouter(WithWidth)
  dispatch={[Function]}
  params={
    Object {
      "username": "123",
    }
  }
  router={
    Object {
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "isActive": [Function],
      "push": [Function],
      "replace": [Function],
      "setRouteLeaveHook": [Function],
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  users={
    Array [
      Object {
        "email": "manolo@gmail.com",
        "gender": "male",
        "name": Object {
          "first": "Manolo",
          "last": "lama",
          "title": "mr",
        },
        "username": "mano",
      },
    ]
  }
  width={3}
/>
`;

exports[`<UserListContainer /> should render with default props 1`] = `
<UserList
  params={
    Object {
      "username": "123",
    }
  }
  router={
    Object {
      "push": [Function],
    }
  }
  showUserProfile={[Function]}
  username="123"
  users={Array []}
  width={3}
/>
`;
